Plans 2023: biannual toggle foundation, behind feature flag

chriskmnds opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/wp-calypso/issues/74260

Part of addressing pdgrnI-28f-p2

## Proposed Changes

- Introduces the 1-year/2-year toggle behind a feature flag `plans/biannual-toggle`
    - The price description for biannual plans reads _"per month, [PRICE] billed every two years"_
- No discounted prices shown in main price displays (addressed in separate PR)

### Media

<img width="650" alt="Screenshot 2023-03-10 at 2 03 49 PM" src="https://user-images.githubusercontent.com/1705499/224311442-9b901b8f-3340-4171-a23a-2b458102944a.png">

<img width="650" alt="Screenshot 2023-03-10 at 2 21 22 PM" src="https://user-images.githubusercontent.com/1705499/224314765-dd7cf5e0-7504-46cc-bddd-10a414425b57.png">


On `/plans/2yearly/[SITE_URL]` "Plans" should be selected in navbar:

<img width="650" alt="Screenshot 2023-03-10 at 2 13 01 PM" src="https://user-images.githubusercontent.com/1705499/224313529-3e887c10-6d85-4a23-8ceb-755448ffac7b.png">

### Notes

2-year plans are also selectable/visible if visiting `/plans/2yearly/[SITE_URL]` without feature flag (same with `trunk`).

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- go to `/plans/[SITE_URL]` with the feature flag `&flags=plans/biannual-toggle`
- confirm 1-year/2-year toggle shows, as in the media
- ensure prices are correct through to the checkout page
- **(NEW)** visit `/plans/2yearly/[SITE]` and ensure sums are correct in the Price descriptions (should be `price * 24`)
- **(NEW)** visit `/plans/2yearly/[SITE]` with a site on a paid plan (e.g. Business) and ensure the `Your plan` badge is visible, and the button sends to `Manage plan`
- ensure on visiting `/plans/2yearly/[SITE_URL]` "Plans" is selected in navbar, as in media (with/out feature flag)
- disable the feature flag and ensure the monthly toggle shows instead

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

chriskmnds committed:
billing period, pricing desc

chriskmnds committed:
move switch behind feature flag plans/biannual-toggle

chriskmnds committed:
move biannual-toggle behind feature flag

chriskmnds committed:
update navbar to highlight Plans

chriskmnds commented on client/my-sites/plan-features-2023-grid/billing-timeframe.jsx:
```diff
@@ -25,9 +26,17 @@ function getPerMonthDescription( {
-		return translate( 'per month, %(annualPriceText)s billed annually', {
-			args: { annualPriceText },
-		} );
+		if ( billingPeriod === 365 ) {
```
grab from `calypso-products/constants/terms.ts`

southp replied:
TIL our billing period is defined this way 😆 
Does that mean we can also import the constants and use them here? i.e. PLAN_ANNUAL_PERIOD

chriskmnds replied:
Yep. Updated to use the constants instead.

There's also the alternative to refactor the `isYearly` (etc.) selectors to accept `plans` - apparently we have those, but they are typed for `products`, which will need some refactoring. Maybe for another time? (if we ever get to) 🤷‍♂️

southp commented on client/my-sites/plan-features-2023-grid/billing-timeframe.jsx:
```diff
@@ -25,9 +26,17 @@ function getPerMonthDescription( {
+			} );
+		}
+
+		if ( billingPeriod === 730 ) {
```
The same as above, could it be `PLAN_BIENNIAL_PERIOD`?

southp commented on packages/data-stores/src/plans/types.ts:
```diff
@@ -57,7 +57,7 @@ export interface PricedAPIPlan {
 	product_slug: StorePlanSlug;
 	product_name_short: string;
-	bill_period: -1 | 31 | 365;
+	bill_period: -1 | 31 | 365 | 730;
```
I'm also wondering if we can somehow unify this with the [PERIOD_LIST](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-products/src/constants/terms.ts#L13) here 🤔 
If not, could we replace 31, 365, 730 by the same set of constants?

chriskmnds replied:
> I'm also wondering if we can somehow unify this with the [PERIOD_LIST](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-products/src/constants/terms.ts#L13) here 🤔

I am not sure if we can do that right now. `data-stores` is on [NPM](https://www.npmjs.com/package/@automattic/data-stores), while `calypso-products` hasn't been published yet. We can use `workspace` dependencies, which (unless I am mistaken) will pick the dependency off the Calypso repo, but that will probably create issues if a new version is pushed to NPM. I think..

but also wondering if thinking in terms of coupling `data-stores/plans` with `calypso-prodicts/plans` would be ideal long term. I'd think the former should come to replace the latter entirely at some point i.e. `data-stores/plans` being enough, so pros and cons I guess

> If not, could we replace 31, 365, 730 by the same set of constants?

Yes, this should be fine for now. There are places in the package where these values are being used. Happy to introduce the constants and use them for this instance, but maybe we can handle a wider refactor in some other PR. We may need to deep dive a little more to test this PR as a bigger `data-stores` refactor 🤔

southp replied:
> I am not sure if we can do that right now. data-stores is on [NPM](https://www.npmjs.com/package/@automattic/data-stores), while calypso-products hasn't been published yet. We can use workspace dependencies, which (unless I am mistaken) will pick the dependency off the Calypso repo, but that will probably create issues if a new version is pushed to NPM. I think..

Agreed. Introducing dependency between `data-stores/plans` and `calypso-products/plans` sounds muddy, since they feel roughly at the same level of the whole dependency hierarchy. Let's not do that to limit the scope of this PR.

> maybe we can handle a wider refactor in some other PR. We may need to deep dive a little more to test this PR as a bigger data-stores refactor 🤔

Also agreed. Let's keep the scope of this PR tangible :)

chriskmnds replied:
Ok - added the constants and updated the type here. We can maybe create a low-priority follow-up to clean up the package and use these elsewhere.

southp reviewed (commented):
I've tested the following scenarios and it works great:

1. Go through /start and see the 1yr/2yr toggle.
2. After toggling to the 2yr, choose a plan and see the 2yr plan is added to the cart in the checkout page.
3. In /plans, see the 1yr/2yr toggle.
4. Visting /plans/2yearly and /plans/yearly leads me to the correct term of plans.
5. Without the feature flag enabled, everything stays the same as how the production works.

Code-wise I only have some questions regarding reusing the defined constants as described in the inline comments :)

chriskmnds commented:
Spotted an issue while testing. The "current plan" functionalities in the grid need to be updated for the 2-year variants. This also happens in trunk if someone visits `/plans/2yearly/[SITE]`, but obviously surfaces now that we'll give a toggle for users to view.

In `/plans/yearly/[SITE]` or `/plans/monthly/[SITE]` with site on Business:

<img width="700" alt="Screenshot 2023-03-14 at 4 04 58 PM" src="https://user-images.githubusercontent.com/1705499/225027470-6b141969-71e9-4a3c-8673-4ef1484d2fce.png">

In `/plans/2yearly/[SITE]` with site on Business:

<img width="700" alt="Screenshot 2023-03-14 at 4 05 14 PM" src="https://user-images.githubusercontent.com/1705499/225027498-813e0fa8-d027-4beb-8fba-dd77e24b0a8d.png">

chriskmnds commented:
Thanks for testing @southp . Moving this back to "in progress" due to the issue mentioned [here](https://github.com/Automattic/wp-calypso/pull/74206#issuecomment-1468176554). will resurface this for reviews again later.

southp commented:
Great catch! Thanks for paying so much attention to details, which is something I really need to learn from you. I'll be more than happy to re-review when it's ready :)

chriskmnds commented:
Spotted one more. With a site on paid/Business plan, the sums in the description don't match in `/plans/2yearly/[SITE]` e.g. should be `564`, not `282`:

<img width="700" alt="Screenshot 2023-03-15 at 1 53 39 PM" src="https://user-images.githubusercontent.com/1705499/225301818-c3bd7df3-4042-4703-a8ca-f9c2fc9bb44a.png">

chriskmnds committed:
fix maybe-discounted price sums in desc. refactor rawPriceAnnual -> maybeDiscountedPriceFullTerm

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -1060,9 +1063,9 @@ const ConnectedPlanFeatures2023Grid = connect(
 			}
 
-			const rawPriceAnnual =
+			const maybeDiscountedPriceFullTerm =
```
For lack of a better term. Just the existence of a bad/strange name is probably a signal of underlying complexities that need resolution. Technically, we have a "maybe" discounted price for the plan's full term (annual or biannual) that may or may not differ from `rawPrice`. 

We can tweak the naming a bit (the "term" can be more specific to annual/biannual), but that aside, it might be better to leave the dragon lurking and refactor more holistically at some point.

chriskmnds committed:
your plan for biannual grid

chriskmnds committed:
cleanup annualPrice.. -> fullTermPrice..

chriskmnds committed:
use billing period constants

chriskmnds committed:
refactor maybeDiscountedPriceFullTerm -> maybeDiscountedFullTermPrice

chriskmnds committed:
ts billing-timeframe and cleanup

chriskmnds commented on packages/calypso-products/src/get-plan-term-variant.ts:
```diff
@@ -0,0 +1,26 @@
+ * @param {typeof TERMS_LIST[ number ]} planTerm The term to find a variant for
+ * @returns {PlanSlug} Slug of the term-variant plan
+ */
+export function getPlanSlugForTermVariant(
```
We have a few (incomplete) ways to do similar right now e.g. per term selectors `getYearlyFromMonthly` (etc.), or even `getAnnualSlug` props on monthly product definitions in the product lists). I think a canonical/polymorphic form should serve things better for any sort of plan conversion that may come up. If we leave this out there, things may gradually converge here (rather than continue with the old). 

There's a similar utility for products ([getProductYearlyVariant](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-products/src/get-product-yearly-variant.ts)) but also built a little more restrictively.

~Happy with either/or~ (well not exactly, but I won't grudge :D). If we keep this, we can add some tests to guard.

southp replied:
The introduction of 2-yearly and 3-yearly plans basically means we need to generalize the code regarding the term eventually, and this looks like a great start :) 

We can surely add test coverage as a separate PR to keep the scope of this PR.

chriskmnds replied:
> We can surely add test coverage as a separate PR to keep the scope of this PR.

Yes, let's do that to keep things moving. 🙌 We can come back to this later this/next week.

New issue: https://github.com/Automattic/wp-calypso/issues/74640

chriskmnds commented on client/my-sites/plan-features-2023-grid/billing-timeframe.tsx:
```diff
@@ -0,0 +1,93 @@
+interface Props {
+	planName: string;
+	billingTimeframe: TranslateResult;
+	billingPeriod: number | null | undefined;
+	rawPrice: number | null;
+	maybeDiscountedFullTermPrice: number | null;
+	annualPricePerMonth: number | null;
```
These optional nulls are kinda unfortunate, stemming from how the underlying selectors/structures behave. I think one alternative, here, would be to condition where the component is imported/used, and instead set these to just the expected types (`number`). Not sure if worth/better though.

chriskmnds committed:
billing period constants in data-stores/plans

chriskmnds committed:
add period_list array

chriskmnds commented:
@southp this is ready for another review/testing. Updated to address your feedback and the issues mentioned above:

- For https://github.com/Automattic/wp-calypso/pull/74206#issuecomment-1468176554, we now show "your plan" for the biannual variants of the site's plan.
- For https://github.com/Automattic/wp-calypso/pull/74206#issuecomment-1469872585, we now show the correct 2-year projected price (this was only an issue on discounted prices - so to test, go to `/plans/2yearly/[SITE]` with a site on a paid/Business plan to see some discounts)

chriskmnds commented:
> For https://github.com/Automattic/wp-calypso/pull/74206#issuecomment-1469872585, we now show the correct 2-year projected price (this was only an issue on discounted prices - so to test, go to /plans/2yearly/[SITE] with a site on a paid/Business plan to see some discounts)

Sorry, this is confusing/wrong. Updated the comment. guess I got mixed up on wrangling two PRs 🤣 

Also updated testing instructions with `(NEW)`, so just refer to those 🙏 

@southp

southp reviewed (approved):
I've gone through the combinations of having/not-having the feature flag cross with/without the currency-based first-year discount, and they all looks good to me. e.g.
TWD:
<img width="1381" alt="截圖 2023-03-20 上午11 50 26" src="https://user-images.githubusercontent.com/1842898/226268552-b7e1a5fe-30e2-4b0b-a33c-f2ab9792ebfa.png">

PHP:
<img width="1263" alt="截圖 2023-03-20 上午11 48 35" src="https://user-images.githubusercontent.com/1842898/226268622-206f0a3e-55d1-4206-abdf-26f7d6c2697d.png">

Along the way testing this PR I've gathered several other flaws that's been there for ages though. However, they are unrelated to this PR and I believe we will have a better foundation for addressing them with this PR shipped first :)

southp commented:
There are two follow-up issues I've found after approval. However, they don't change my stance that this PR should be shipped as a better foundation first. So please feel free to handle them as follow-up PRs :) 

1. Visiting /plans, the 1y/2y toggle is missing when I'm on an annual plan. Currently it is hide because it's problematic to downgrade from an annual term to a monthly one. Now that we are showing a longer term option, it makes sense to show the toggle . It would also open up the question regarding what we want to show for the monthly plan owner, though. We can discuss that further separately. 
1. Instead of showing "manage plan", providing action for upgrading to a longer term would be better. That's also how it works when someone is on a monthly plan viewing the annual plans page:
<img width="1274" alt="截圖 2023-03-20 下午1 55 08" src="https://user-images.githubusercontent.com/1842898/226276812-240fa16f-aecf-4581-b6ba-7b8b133dfbbc.png">

southp commented:
Sorry, one more thing: the 3yr case somehow reads "/ month" instead of "per month" like the rest. I'm also good with us addressing it as a follow-up :)
<img width="1218" alt="image" src="https://user-images.githubusercontent.com/1842898/226287432-ce8ec309-d869-4b7a-92b2-06e8ea6ab0e2.png">

chriskmnds commented:
Thanks, @southp. I agree - since we are fine with the code changes, then let's merge and tune things in follow-ups. I think it gives us a better foundation to build from. I moved your suggestions into issues to discuss and tackle separately (https://github.com/Automattic/wp-calypso/issues/74636, https://github.com/Automattic/wp-calypso/issues/74637, https://github.com/Automattic/wp-calypso/issues/74638).

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7878421).
